### PR TITLE
[fix][broker] Fix potential can't remove producer/reader reference

### PR DESCRIPTION
**Pull request by Technoboy-**
### Motivation

If the `producer`/`reader` close with the exception, there may exist references that can't remove from the list.

### Documentation

- [x] `doc-not-needed` 
(Please explain why)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java [hunk 1]
```diff
@@ -121,15 +121,17 @@ private String getEventKey(PulsarEvent event) {
 
         @Override
         public void close() throws IOException {
-            this.producer.close();
-            systemTopicClient.getWriters().remove(TopicPolicyWriter.this);
+            try {
+                closeAsync().get();
+            } catch (Exception e) {
+                throw new IOException(e);
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java:127]: this seems wrong. I'd suggest something similar as what is used in BrokerService.close: https://github.com/apache/pulsar/blob/a7f1a5657782ff7ac91a89afdd49568d74d111f5/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L713-L721

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java [hunk 2]
```diff
@@ -184,15 +186,17 @@ public CompletableFuture<Boolean> hasMoreEventsAsync() {
 
         @Override
         public void close() throws IOException {
-            this.reader.close();
-            systemTopic.getReaders().remove(TopicPolicyReader.this);
+            try {
+                closeAsync().get();
+            } catch (Exception e) {
+                throw new IOException(e);
+            }
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java:193]: this seems wrong. I'd suggest something similar as what is used in BrokerService.close: https://github.com/apache/pulsar/blob/a7f1a5657782ff7ac91a89afdd49568d74d111f5/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L713-L721

**Review by lhotari — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java (hunk 1 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java:127]: +1

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java (hunk 2 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java:193]: +1

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java]: fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java (hunk 2 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java]: fixed

**Review by lhotari — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java [hunk 3]
```diff
@@ -184,15 +194,23 @@ public CompletableFuture<Boolean> hasMoreEventsAsync() {
...
+                    throw new PulsarServerException(e.getCause());
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
         }
 
         @Override
         public CompletableFuture<Void> closeAsync() {
-            return reader.closeAsync().thenCompose(v -> {
```
Comment (HQebupt) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java:193]: This line is key point. The previous `thenCompose` is replaced by `thenApply` just to save `CompletableFuture`. Right?

**Review by HQebupt — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java (hunk 3 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java:193]: yes.

**Review by mattisonchao — APPROVED** [review]

**Pull request merged**